File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        int maxLen = 0;
        unordered_set<int>st;
        int i = 0, j = 0;
        while(j < n){
            if(st.count(s[j]) == 0) st.insert(s[j]);
            else{
                while(st.count(s[j])){ // try to reduce from the i side
                    st.erase(s[i]);
                    i++;
                }
                st.insert(s[j]);
            }
            maxLen = max(maxLen, j-i+1); 
            j++;
        }


        return maxLen;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int maxLen = 0;
        HashSet<Character>st = new HashSet<>();
        int i = 0;
        for(int r=0; r<n; r++){
            char ch = s.charAt(r);
            while(st.contains(ch)){
                st.remove(s.charAt(i));
                i++;
            }
            st.add(ch);
            maxLen = Math.max(maxLen, st.size());
        }  
        return maxLen; 
    }
}